RFE: Support for automatic creation of https monitors with SNI

#### Title
Support for automatic creation of https monitors with SNI

#### Description

At time of this writing it is required for customers to create https monitors in order to health check an application that is hosting several domain names. This is the typical case of sending traffic to an ingress controller (2-tier deployments).

This requires:

- the creation of a server TLS profile
- the creation of the monitor itself

as follows:

```
ltm profile server-ssl serverssl.www.mc-nginx.com {
    app-service none
    defaults-from serverssl
    server-name www.mc-nginx.com
}
ltm monitor https www.mc-nginx.com {
    defaults-from https
    send "GET / HTTP/1.1\r\nHost: www.mc-nginx.com\r\nConnection: close\r\n\r\n"
    ssl-profile /Common/serverssl.www.mc-nginx.com
}
```

#### Actual Problem

Customers cannot automate the monitoring for this typical use case because the monitors need to be pre-created

#### Solution Proposed

When indicating an https monitor in the manifest, this should be by default SNI enabled

#### Additional context

In order for SNI enabled monitors to work in BIG-IP Classic it is necessary to enable the following DB key

```
tmsh modify sys db bigd.tmm value enable
```

See https://my.f5.com/manage/s/article/K85210713 for more details. Given that in-TMM monitoring has been affected by issues in the past and has an impact in data plane's performance. I would recommend that https monitors with SNI (which require in-TMM right now) have a different name than https in the manifests, for example: https-sni

Note that one of the TLS server profiles of the monitors will need to have ticked the option 

```
    sni-default true
```
otherwise AS3 complains (I believe it shouldn´t in this case), maybe an AS3 bug should be raised?

"2024/01/17 14:26:38 [ERROR] [AS3] Raw response from Big-IP: map[code:422 declaration:HTML Tag-like Content in the Request URL/Body results:[map[code:422 host:localhost message:declaration failed response:0107180b:3: Virtual Server /mc-twotier/Shared/crd_10_1_10_115_443 has more than one clientssl/serverssl profile but none of them is default for SNI. runTime:3097 tenant:mc-twotier]]]"

--- a/config/apis/cis/v1/types.go
+++ b/config/apis/cis/v1/types.go
@@ -148,6 +148,7 @@ type Monitor struct {
 	TargetPort int32  `json:"targetPort"`
 	Name       string `json:"name,omitempty"`
 	Reference  string `json:"reference,omitempty"`
+	SSLProfile string `json:"sslProfile,omitempty"`
 }
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object

--- a/docs/RELEASE-NOTES.rst
+++ b/docs/RELEASE-NOTES.rst
@@ -10,6 +10,7 @@ Added Functionality
     * Multi Cluster
     * CRD
       * Support for mix of k8s Secret and bigip reference in TLSProfile.
+      * Support for setting sslProfile with https monitor in virtualServer and nextgen routes
 
 Bug Fixes
 ````````````

--- a/docs/config_examples/customResource/CustomResource.md
+++ b/docs/config_examples/customResource/CustomResource.md
@@ -137,16 +137,17 @@ Note: **monitors** take priority over **monitor** if both are provided in VS spe
 
 **Health Monitor**
 
-| PARAMETER | TYPE | REQUIRED | DEFAULT | DESCRIPTION                                                                                                                         |
-| ------ | ------ | ------ | ------ |-------------------------------------------------------------------------------------------------------------------------------------|
-| type | String | Required | NA | http, https or tcp                                                                                                                  |
-| send | String | Required | “GET /rn” | HTTP request string to send.                                                                                                        |
-| recv | String | Optional | NA | String or RegEx pattern to match in first 5,120 bytes of backend response.                                                          |
-| interval | Int | Required | 5 | Seconds between health queries                                                                                                      |
-| timeout | Int | Optional | 16 | Seconds before query fails                                                                                                          |
-| targetPort | Int | Optional | 0 | port (if any) monitor should probe ,if 0 (default) then pool member port is used.Translates to "Alias Service Port" on BIG-IP pool. |
-| name | String | Required | NA | Reference to health monitor name existing on bigip                                                                                  |
-| reference | String  | Required | NA | Value should be bigip for referencing custom monitor on bigip                                                                       |
+| PARAMETER  | TYPE   | REQUIRED | DEFAULT   | DESCRIPTION                                                                                                                         |
+|------------|--------|----------|-----------|-------------------------------------------------------------------------------------------------------------------------------------|
+| type       | String | Required | NA        | http, https or tcp                                                                                                                  |
+| send       | String | Required | “GET /rn” | HTTP request string to send.                                                                                                        |
+| recv       | String | Optional | NA        | String or RegEx pattern to match in first 5,120 bytes of backend response.                                                          |
+| interval   | Int    | Required | 5         | Seconds between health queries                                                                                                      |
+| timeout    | Int    | Optional | 16        | Seconds before query fails                                                                                                          |
+| targetPort | Int    | Optional | 0         | port (if any) monitor should probe ,if 0 (default) then pool member port is used.Translates to "Alias Service Port" on BIG-IP pool. |
+| name       | String | Required | NA        | Reference to health monitor name existing on bigip                                                                                  |
+| reference  | String | Required | NA        | Value should be bigip for referencing custom monitor on bigip                                                                       |
+| sslProfile | String | Optional | NA        | sslProfile to attach to custom https monitor created on BIGIP.Applicable only for type "https" monitor                              |
 
 **TCP Profile Components**
 

--- a/docs/config_examples/customResource/VirtualServerWithTLSProfile/tls-with-health-monitor/virtualserver.yml
+++ b/docs/config_examples/customResource/VirtualServerWithTLSProfile/tls-with-health-monitor/virtualserver.yml
@@ -15,6 +15,7 @@ spec:
         send: /
         timeout: 10
         type: https
+        sslProfile: /Common/serverssl
       path: /bar
       service: svc-3
       servicePort: 80

--- a/docs/config_examples/customResourceDefinitions/incubator/customresourcedefinitions.yml
+++ b/docs/config_examples/customResourceDefinitions/incubator/customresourcedefinitions.yml
@@ -174,6 +174,9 @@ spec:
                             type: string
                           recv:
                             type: string
+                          sslProfile:
+                            type: string
+                            pattern: '^\/([A-z0-9-_+]+\/)+([A-z0-9]+\/?)*$'
                     reference:
                       type: string
                       enum: [ bigip, service ]
@@ -265,6 +268,9 @@ spec:
                           reference:
                             type: string
                             enum: [bigip]
+                          sslProfile:
+                            type: string
+                            pattern: '^\/([A-z0-9-_+]+\/)+([A-z0-9]+\/?)*$'
                       monitors:
                         type: array
                         items:
@@ -289,6 +295,9 @@ spec:
                             reference:
                               type: string
                               enum: [bigip]
+                            sslProfile:
+                              type: string
+                              pattern: '^\/([A-z0-9-_+]+\/)+([A-z0-9]+\/?)*$'
                       minimumMonitors:
                         x-kubernetes-int-or-string: true
                         anyOf:

--- a/docs/config_examples/next-gen-routes/routes/route-with-target-port-health-monitor.yaml
+++ b/docs/config_examples/next-gen-routes/routes/route-with-target-port-health-monitor.yaml
@@ -10,6 +10,8 @@ metadata:
           "interval": 5,
           "timeout": 10,
           "targetPort": 8080,
+          "type": "https",
+          "sslProfile": "/Common/serverssl"
         }
       ]
   labels:

--- a/pkg/controller/backend.go
+++ b/pkg/controller/backend.go
@@ -1594,6 +1594,9 @@ func createMonitorDecl(cfg *ResourceConfig, sharedApp as3Application) {
 			}
 			monitor.Send = v.Send
 			monitor.TimeUnitilUp = v.TimeUntilUp
+			if v.SSLProfile != "" {
+				monitor.ClientTLS = as3ResourcePointer{BigIP: fmt.Sprintf("%v", v.SSLProfile)}
+			}
 		case "tcp", "udp":
 			adaptiveFalse := false
 			monitor.Adaptive = &adaptiveFalse

--- a/pkg/controller/nativeResourceWorker.go
+++ b/pkg/controller/nativeResourceWorker.go
@@ -545,6 +545,7 @@ func (ctlr *Controller) prepareResourceConfigFromRoute(
 						Timeout:    hm.Timeout,
 						Path:       hm.Path,
 						TargetPort: hm.TargetPort,
+						SSLProfile: hm.SSLProfile,
 					}
 					rsCfg.Monitors = append(
 						rsCfg.Monitors,

--- a/pkg/controller/resourceConfig.go
+++ b/pkg/controller/resourceConfig.go
@@ -798,6 +798,7 @@ func (ctlr *Controller) createVirtualServerMonitor(monitor cisapiv1.Monitor, poo
 				Recv:       monitor.Recv,
 				Timeout:    monitor.Timeout,
 				TargetPort: monitor.TargetPort,
+				SSLProfile: monitor.SSLProfile,
 			}
 			rsCfg.Monitors = append(rsCfg.Monitors, monitor)
 		}

--- a/pkg/controller/types.go
+++ b/pkg/controller/types.go
@@ -470,6 +470,7 @@ type (
 		TargetPort  int32  `json:"targetPort,omitempty"`
 		Path        string `json:"path,omitempty"`
 		TimeUntilUp *int   `json:"timeUntilUp,omitempty"`
+		SSLProfile  string `json:"sslProfile,omitempty"`
 	}
 	MonitorName struct {
 		Name string `json:"name"`
@@ -1042,19 +1043,20 @@ type (
 	// - Monitor_HTTP
 	// - Monitor_HTTPS
 	as3Monitor struct {
-		Class             string  `json:"class,omitempty"`
-		Interval          int     `json:"interval,omitempty"`
-		MonitorType       string  `json:"monitorType,omitempty"`
-		TargetAddress     *string `json:"targetAddress,omitempty"`
-		Timeout           int     `json:"timeout,omitempty"`
-		TimeUnitilUp      *int    `json:"timeUntilUp,omitempty"`
-		Adaptive          *bool   `json:"adaptive,omitempty"`
-		Dscp              *int    `json:"dscp,omitempty"`
-		Receive           string  `json:"receive"`
-		Send              string  `json:"send"`
-		TargetPort        int32   `json:"targetPort,omitempty"`
-		ClientCertificate string  `json:"clientCertificate,omitempty"`
-		Ciphers           string  `json:"ciphers,omitempty"`
+		Class             string             `json:"class,omitempty"`
+		Interval          int                `json:"interval,omitempty"`
+		MonitorType       string             `json:"monitorType,omitempty"`
+		TargetAddress     *string            `json:"targetAddress,omitempty"`
+		Timeout           int                `json:"timeout,omitempty"`
+		TimeUnitilUp      *int               `json:"timeUntilUp,omitempty"`
+		Adaptive          *bool              `json:"adaptive,omitempty"`
+		Dscp              *int               `json:"dscp,omitempty"`
+		Receive           string             `json:"receive"`
+		Send              string             `json:"send"`
+		TargetPort        int32              `json:"targetPort,omitempty"`
+		ClientCertificate string             `json:"clientCertificate,omitempty"`
+		Ciphers           string             `json:"ciphers,omitempty"`
+		ClientTLS         as3ResourcePointer `json:"clientTLS,omitempty"`
 	}
 
 	// as3CABundle maps to CA_Bundle in AS3 Resources

